packages/react-dom/src/__tests__/ReactDOMFiber-test.js

```diff
@@ -1280,4 +1280,36 @@ describe('ReactDOMFiber', () => {
     );
     expect(didCallOnChange).toBe(true);
   });
+
+  it('unmounted legacy roots should never clear newer root content from a container', () => {
+    const ref = React.createRef();
+
+    function OldApp() {
+      const hideOnFocus = () => {
+        // This app unmounts itself inside of a focus event.
+        ReactDOM.unmountComponentAtNode(container);
+      };
+
+      return (
+        <button onFocus={hideOnFocus} ref={ref}>
+          old
+        </button>
+      );
+    }
+
+    function NewApp() {
+      return <button ref={ref}>new</button>;
+    }
+
+    ReactDOM.render(<OldApp />, container);
+    ref.current.focus();
+
+    ReactDOM.render(<NewApp />, container);
+
+    // Calling focus again will flush previously scheduled discerete work for the old root-
```

Comment:
Something is off here because there's no concept of "discrete work" for legacy mode (and I'm removing it for CM too). At least there shouldn't be. Maybe for passive effects but that doesn't kick in here.

There's only Sync work in legacy mode.

What do you mean by discrete work here? It might that something is misnamed in a method or something.

Reply:
@sebmarkbage This comment was based on [Dominic's comment above](https://github.com/facebook/react/pull/18792#issuecomment-622052111).

Reply:
@sebmarkbage Upon calling `scheduleUpdateOnFiber`, you add an entry to `rootsWithPendingDiscreteUpdates` which is the problem in this issue. I'd assume your work of removing the concept of discrete work didn't include this particulat code-path?

Reply:
I haven't landed that part yet. Still blocked on some stuff.

Concretely I'm worried that we've accidentally landed "discrete" semantics in legacy mode at some point. I.e. some subtle breaking change.